### PR TITLE
Lazy completion check

### PR DESCRIPTION
**Pull request by TMaYaD**
## Description
This tries to improvise luigi scheduler by addressing bottlenecks

## Motivation and Context

When luigi worker starts it builds a dependency graph and checks for completion of all tasks.
If we start **w** workers with **t** tasks each, we are instantly checking **(w * t)** tasks for completion which generally are `output().exists()` calls. In a typical use case **w** is expected to increase linearly with **t** so that we could parallelise efficiently. Instead, we end up with a bottleneck with O(n^2) calls.

Changing completion check to lazy addresses this.
- First worker builds the entire graph with with `PENDING` status and submits it to the scheduler
- Scheduler merges it with it's existing graph and returns with a list of tasks to complete
- Worker checks each task it receives for completion and proceeds to try and run it.
  - If task is already completed, worker informs scheduler about the same and gets next task
  - If task is not completed, worker proceeds to run it and informs the scheduler about the result(`DONE`/`FAIL`)
- Scheduler as usual makes sure only one worker is working on a task so there are no collisions even while we mark `DONE` tasks as `PENDING`

However, starting with the PENDING status for tasks results in already DONE/FAIL tasks to be (wrongly) transitioned to PENDING when a new worker starts and submits it's graph. Hence a `NEW` task state is introduced. Which is a pseudo state and replaced by scheduler to existing task state or `PENDING` as appropriate. This change makes the workers run tasks correctly. 

One final issue remaining is that now the task status is not synchronised between worker and scheduler and as a result the execution summary is off. We remedy this by returning the modified task status(and other info) from scheduler to worker and updating the worker's copy of the task.

## Have you tested this? If so, how?
This is tested manually using `examples.touch_n.NTask` 
The automated test suite could not be run.

### CODE REVIEW

File: luigi/scheduler.py [hunk 1]
```diff
@@ -826,9 +833,12 @@ def add_task(self, task_id=None, status=PENDING, runnable=True,
...
                     batch_task.expl = expl
 
-        if not (task.status in (RUNNING, BATCH_RUNNING) and (status not in (DONE, FAILED, RUNNING) or task.worker_running != worker_id)) or new_deps:
+        task_is_not_running = task.status not in (RUNNING, BATCH_RUNNING)
+        task_started_a_run = status in (DONE, FAILED, RUNNING)
+        running_on_this_worker = task.worker_running == worker_id
+        if task_is_not_running or (task_started_a_run and running_on_this_worker) or new_deps:
             # don't allow re-scheduling of task while it is running, it must either fail or succeed on the worker actually running it
-            if status == PENDING or status != task.status:
+            if status != task.status or status == PENDING:
```
Comment (abhik) [inline luigi/scheduler.py]: Why this change? There are other changes in this PR that make the code more readable but will make it more difficult to merge upstream changes.

**Comment by TMaYaD** [comment]
Good call on Dynamic dependencies. I'll do that and test.

File: luigi/scheduler.py [hunk 2]
```diff
@@ -520,6 +520,13 @@ def set_batch_running(self, task, batch_id, worker_id):
         task.time_running = time.time()
 
     def set_status(self, task, new_status, config=None):
+        logger.debug("Transitioning Task {} from {} to {}.".format(task.id, task.status, new_status))
```
Comment (ramyala) [inline luigi/scheduler.py]: Add a comment on why you have to override the input status

File: luigi/scheduler.py [hunk 3]
```diff
@@ -691,11 +698,11 @@ def dump(self):
 
     @rpc_method()
     def prune(self):
-        logger.info("Starting pruning of task graph")
+        logger.debug("Starting pruning of task graph")
```
Comment (ramyala) [inline luigi/scheduler.py]: get rid of debug logs which aren't required in production

File: luigi/scheduler.py [hunk 4]
```diff
@@ -826,9 +833,12 @@ def add_task(self, task_id=None, status=PENDING, runnable=True,
                 for batch_task in self._state.get_batch_running_tasks(task.batch_id):
                     batch_task.expl = expl
 
-        if not (task.status in (RUNNING, BATCH_RUNNING) and (status not in (DONE, FAILED, RUNNING) or task.worker_running != worker_id)) or new_deps:
+        task_is_not_running = task.status not in (RUNNING, BATCH_RUNNING)
+        task_started_a_run = status in (DONE, FAILED, RUNNING)
+        running_on_this_worker = task.worker_running == worker_id
+        if task_is_not_running or (task_started_a_run and running_on_this_worker) or new_deps:
```
Comment (ramyala) [inline luigi/scheduler.py]: revert these changes back

File: luigi/scheduler.py [hunk 5]
```diff
@@ -885,6 +895,25 @@ def add_task(self, task_id=None, status=PENDING, runnable=True,
             self._state.get_worker(worker_id).tasks.add(task)
             task.runnable = runnable
 
+        return {
```
Comment (ramyala) [inline luigi/scheduler.py]: Replicating task structure isn't ideal. Return a deepcopy of task object.

File: luigi/worker.py [hunk 6]
```diff
@@ -180,16 +180,12 @@ def run(self):
...
-                if self.task.complete():
-                    status = DONE
-                else:
-                    status = FAILED
-                    expl = 'Task is an external data dependency ' \
-                        'and data does not exist (yet?).'
+            if self.task.complete():
+                status = DONE
+            elif _is_external(self.task):
+                status = FAILED
```
Comment (ramyala) [inline luigi/worker.py]: does the TODO in original code and issue #814 get resolved with your change?

File: luigi/worker.py [hunk 7]
```diff
@@ -483,22 +479,23 @@ def _add_task(self, *args, **kwargs):
         Call ``self._scheduler.add_task``, but store the values too so we can
         implement :py:func:`luigi.execution_summary.summary`.
         """
-        task_id = kwargs['task_id']
-        status = kwargs['status']
-        runnable = kwargs['runnable']
+        task_info = self._scheduler.add_task(*args, **kwargs)
+        task_id = task_info['task']['task_id']
```
Comment (ramyala) [inline luigi/worker.py]: change interface to query from task object.

File: luigi/scheduler.py (hunk 5 above)
Comment (TMaYaD) [reply luigi/scheduler.py]: This method is accessed by remote workers as an RPC method and requires the return value to be JSON serialisable.

File: luigi/worker.py (hunk 6 above)
Comment (TMaYaD) [reply luigi/worker.py]: yes.

File: luigi/worker.py (hunk 7 above)
Comment (TMaYaD) [reply luigi/worker.py]: See above on why the interface is a dict instead of task object.

File: luigi/scheduler.py [hunk 8]
```diff
@@ -533,7 +543,7 @@ def set_status(self, task, new_status, config=None):
                 self.re_enable(task)
 
             # don't allow workers to override a scheduler disable
-            elif task.scheduler_disable_time is not None and new_status != DISABLED:
+            elif task.scheduler_disable_time is not None:
```
Comment (abhik) [inline luigi/scheduler.py]: Any other consequences of this change? Seems like a worker can re-enable a task before the timeout is over. 

Also, it will be much harder to merge upstream changes in this file because we're altering the behavior of the task's state machine. So, even changes that merge cleanly can introduce a regression.

**Comment by abhik** [comment]
> Instead, we end up with a bottleneck with O(n^2) calls.

I guess I don't have enough context on why we're making this change in the first place. Has this caused a problem in practice? If so, can we solve the problem without changing luigi's scheduler?

**Comment by abhik** [comment]
I know we had a PR open upstream. Did that get merged? What changed today that we're merging this? I'm still uncomfortable with making this change without running any of luigi's tests.

**Comment by abhik** [comment]
Actually, it looks like some tests do fail: https://travis-ci.org/spotify/luigi/jobs/318664621
